array/builder.go

```diff
@@ -0,0 +1,157 @@
+package array
+
+import (
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/apache/arrow/go/arrow/memory"
+)
+
+type StringBuilder struct {
+	builder      *array.BinaryBuilder
+	mem          memory.Allocator
+	value        string
+	length       int
+	capacity     int
+	dataCapacity int
+	refCount     int
+}
+
+func NewStringBuilder(mem memory.Allocator) *StringBuilder {
+	return &StringBuilder{
+		mem:      mem,
+		refCount: 1,
+	}
+}
+func (b *StringBuilder) init() {
+	if b.builder == nil {
+		if b.refCount <= 0 {
+			return
```

Comment:
When would this happen? Should we panic here instead?

Reply:
This can happen when the builder is released and then attempted to be used. I'm not sure a panic is worth it though. This shouldn't really come up as generally the builders shouldn't be used with a reference count of zero. If you want me to add a panic though we can.

Reply:
It seems like if tracking memory is important enough, we want to know about errors like this. But we also don't want to fail the query if it wouldn't otherwise fail. I'm conflict but I guess I'm fine with leaving it as-is.